rcl_yaml_param_parser/src/namespace.c

```diff
@@ -0,0 +1,190 @@
+// Copyright 2018 Apex.AI, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "./impl/namespace.h"
+
+#include "rcutils/strdup.h"
+
+rcutils_ret_t add_name_to_ns(
+  namespace_tracker_t * ns_tracker,
+  const char * name,
+  const namespace_type_t namespace_type,
+  rcutils_allocator_t allocator)
+{
+  char * cur_ns;
+  uint32_t * cur_count;
+  char * sep_str;
+  size_t name_len;
+  size_t ns_len;
+  size_t sep_len;
+  size_t tot_len;
+
+  switch (namespace_type) {
+    case NS_TYPE_NODE:
+      cur_ns = ns_tracker->node_ns;
+      cur_count = &(ns_tracker->num_node_ns);
+      sep_str = NODE_NS_SEPERATOR;
+      break;
+    case NS_TYPE_PARAM:
+      cur_ns = ns_tracker->parameter_ns;
+      cur_count = &(ns_tracker->num_parameter_ns);
+      sep_str = PARAMETER_NS_SEPERATOR;
+      break;
+    default:
+      return RCUTILS_RET_ERROR;
+  }
+
+  /// Add a name to ns
+  if (NULL == name) {
+    return RCUTILS_RET_INVALID_ARGUMENT;
+  }
+  if (0U == *cur_count) {
+    cur_ns = rcutils_strdup(name, allocator);
+    if (NULL == cur_ns) {
+      return RCUTILS_RET_BAD_ALLOC;
+    }
+  } else {
+    ns_len = strlen(cur_ns);
+    name_len = strlen(name);
+    sep_len = strlen(sep_str);
+    // Check the last sep_len characters of the current NS against the separator string.
+    if (strcmp(cur_ns + ns_len - sep_len, sep_str) == 0) {
+      // Current NS already ends with the separator: don't put another separator in.
+      sep_len = 0;
+      sep_str = "";
+    }
+
+    tot_len = ns_len + sep_len + name_len + 1U;
+
+    cur_ns = allocator.reallocate(cur_ns, tot_len, allocator.state);
```

Comment:
@brawner if `allocator.reallocate` fails, `cur_ns` will leak.

Reply:
Moving to https://github.com/ros2/rcl/pull/780